File: Cdb/src/main/python/IasCdb/CdbTxtFiles.py
from pathlib import Path
import os
import shutil

from IasCdb.CdbFolders import CdbFolders, Folders
from IasCdb.TextFileType import FileType, TextFileType
class CdbTxtFiles():

    confFileExtension = ".conf"

    @classmethod
    def from_folder(cls, parent_folder: str):
        """
        Build the the CdbTxtFiles from the contento of the passed folder
        i.e. it guess the type of the CDB (YAML, JSON) by looking at
        the content of the foder
        """
        cdb_type = TextFileType.get_cdb_type(parent_folder)
        if cdb_type is None:
            raise ValueError(f"{parent_folder} is not a parent folder of a CDB")
        return CdbTxtFiles(parent_folder, cdb_type)

    def __init__(self, parent_folder: str, files_type: FileType) -> None:
        self.cdb_parent_folder = parent_folder
        self.files_type = files_type
        self.file_extension = TextFileType.getExtension(files_type)
        self.ias_file_name = "ias" + self.file_extension
        self.templates_file_name = "templates" + self.file_extension
        self.iasios_file_name = "iasios" + self.file_extension
        self.transfer_funs_file_name = "tfs" + self.file_extension

        if self._check_parent_folder(parent_folder):
            self.cdb_parent_folder = parent_folder
        else:
            raise IOError(f"Error accessing {self.cdb_parent_folder}: check permission")

    def _check_parent_folder(self, parent_folder):
        return os.path.exists(parent_folder) and \
            os.path.isdir(parent_folder) \
            and os.access(parent_folder, os.W_OK)

    def get_ias_file_path(self) -> str:
        return os.path.join(CdbFolders.get_subfolder(self.cdb_parent_folder, Folders.ROOT, False), self.ias_file_name)

    def get_supervisor_file_path(self, supervisor_id) -> str:
        return os.path.join(CdbFolders.get_subfolder(self.cdb_parent_folder, Folders.SUPERVISOR, False), f"{supervisor_id}{self.file_extension}")

    def get_dasu_file_path(self, dasu_id: str) -> str:
        return os.path.join(CdbFolders.get_subfolder(self.cdb_parent_folder, Folders.DASU, False), f"{dasu_id}{self.file_extension}")

    def get_asce_file_path(self, asce_id: str) -> str:
        return os.path.join(CdbFolders.get_subfolder(self.cdb_parent_folder, Folders.ASCE, False), f"{asce_id}{self.file_extension}")

    def get_iasio_file_path(self) -> str:
        return os.path.join(CdbFolders.get_subfolder(self.cdb_parent_folder, Folders.IASIO, False), self.iasios_file_name)

    def get_tf_file_path(self) -> str:
        return os.path.join(CdbFolders.get_subfolder(self.cdb_parent_folder, Folders.TF, False), self.transfer_funs_file_name)

    def get_template_file_path(self, template_id: str)-> str:
        return os.path.join(CdbFolders.get_subfolder(self.cdb_parent_folder, Folders.TEMPLATE, False), self.templates_file_name)

    def get_client_file_path(self, client_id):
        if client_id is None or not client_id:
            raise ValueError("Invalid null or empty client ID")
        return os.path.join(CdbFolders.get_subfolder(self.cdb_parent_folder, Folders.CLIENT, False), f"{client_id}{self.confFileExtension}")

    def get_plugin_file_path(self, plugin_id):
        if plugin_id is None or not plugin_id:
            raise ValueError("Invalid null or empty plugin ID")
        return os.path.join(CdbFolders.get_subfolder(self.cdb_parent_folder, Folders.PLUGIN, False), f"{plugin_id}{self.file_extension}")

File: Cdb/src/main/java/org/eso/ias/cdb/structuredtext/CdbTxtFiles.java
package org.eso.ias.cdb.structuredtext;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Objects;

public class CdbTxtFiles implements CdbFiles {

	/**
	 * The extension of file names
	 */
	public final String fileExtension;

	/**
	 * The types of the files in the CDB
	 */
	public final TextFileType filesType;

	/**
	 * The extension of configuration files of the clients
	 */
	public static final String confFileExtension=".conf";
	
	/**
	 * The name of the file containing all the templates
	 */
	public final String iasFileName;
	
	/**
	 * The name of the file containing all the templates
	 */
	public final String templatesFileName;
	
	/**
	 * The name of the file containing all the IASIOs
	 */
	public final String iasiosFileName;
	
	/**
	 * The name of the file containing all the transfer functions
	 */
	public final String transferFunsFileName;
	
	/**
	 * The parent folder of the CDB 
	 * i.e. the folder where we expect to find CDB, CDB/DASU etc.
	 * 
	 * @see CdbFolders
	 */
	public final Path cdbParentFolder;
	
	/**
	 * Constructor
	 * 
	 * @param parentFolder The stringified path of the parent folder of the CDB
	 * @param fType The type of the text files
	 * @throws IOException If the passed folder i snot valid
	 */
	public CdbTxtFiles(String parentFolder, TextFileType fType) throws IOException {
		this(Paths.get(parentFolder), fType);
	}
	
	/**
	 * Constructor
	 * 
	 * @param parentFolder The path of the parent folder of the CDB
     * @param fType The type of the text files
	 * @throws IOException If the passed folder is not valid
	 */
	public CdbTxtFiles(Path parentFolder, TextFileType fType) throws IOException {
		Objects.requireNonNull(parentFolder,"The parent folder can't be null");
		Objects.requireNonNull(fType, "Invalid null file type");
		this.filesType=fType;
		this.fileExtension=fType.ext;
		this.iasFileName="ias"+ fileExtension;
		this.templatesFileName="templates"+ fileExtension;
		this.iasiosFileName="iasios"+ fileExtension;
		this.transferFunsFileName="tfs"+ fileExtension;
		if (checkParentFolder(parentFolder)) {
			this.cdbParentFolder=parentFolder;
		} else {
			throw new IOException("Check folder permission "+parentFolder.toAbsolutePath());
		}
	}
	
	/**
	 * Constructor
	 * 
	 * @param parentFolder The file pointing to the parent folder of the CDB
	 * @param fType The type of the text files
	 * @throws IOException If the passed folder is not valid
	 */
	public CdbTxtFiles(File parentFolder, TextFileType fType) throws IOException {
		this(parentFolder.toPath(), fType);
	}
	
	/**
	 * Check if the passed path is a writable folder.
	 * 
	 * @param p The path to the folder to check
	 * @return <code>true</code> if the path passed all the checks,
	 *         <code>false</code> otherwise
	 */
	private boolean checkParentFolder(Path p) {
		Objects.requireNonNull(p);
		return Files.exists(p) && Files.isDirectory(p) && Files.isWritable(p);
	}

	@Override
	public TextFileType getCdbFileType() {
		return filesType;
	}

	/**
	 * @return The file to store ias configuration
	 * @see CdbFiles#getIasFilePath()
	 */
	@Override
	public Path getIasFilePath()  throws IOException {
		return CdbFolders.getSubfolder(cdbParentFolder, CdbFolders.ROOT).resolve(iasFileName);
	}

	/* (non-Javadoc)
	 * @see org.eso.ias.cdb.json.CdbFiles#getSuperivisorFilePath(java.lang.String)
	 */
	@Override
	public Path getSuperivisorFilePath(String supervisorID)  throws IOException{
		Objects.requireNonNull(supervisorID, "Invalid null supervisor ID");
		if (supervisorID.isEmpty()) {
			throw new IllegalArgumentException("Invalid empty supervisor ID");
		}
		return CdbFolders.getSubfolder(cdbParentFolder, CdbFolders.SUPERVISOR).resolve(supervisorID+ fileExtension);
	}

	/* (non-Javadoc)
	 * @see org.eso.ias.cdb.json.CdbFiles#getDasuFilePath(java.lang.String)
	 */
	@Override
	public Path getDasuFilePath(String dasuID)  throws IOException{
		Objects.requireNonNull(dasuID, "Invalid null DASU ID");
		if (dasuID.isEmpty()) {
			throw new IllegalArgumentException("Invalid empty DASU ID");
		}
		return CdbFolders.getSubfolder(cdbParentFolder, CdbFolders.DASU).resolve(dasuID+ fileExtension);
	}

	/* (non-Javadoc)
	 * @see org.eso.ias.cdb.json.CdbFiles#getAsceFilePath(java.lang.String)
	 */
	@Override
	public Path getAsceFilePath(String asceID)  throws IOException{
		Objects.requireNonNull(asceID, "Invalid null ASCE ID");
		if (asceID.isEmpty()) {
			throw new IllegalArgumentException("Invalid empty ASCE ID");
		}
		return CdbFolders.getSubfolder(cdbParentFolder, CdbFolders.ASCE).resolve(asceID+ fileExtension);
	}

	/* (non-Javadoc)
	 * @see org.eso.ias.cdb.json.CdbFiles#getIasioFilePath(java.lang.String)
	 */
	@Override
	public Path getIasioFilePath(String iasioID)  throws IOException{
		return CdbFolders.getSubfolder(cdbParentFolder, CdbFolders.IASIO).resolve(iasiosFileName);
	}

	/*
	 * (non-Javadoc)
	 * @see org.eso.ias.cdb.json.CdbFiles#getTFFilePath(java.lang.String)
	 */
	@Override
	public Path getTFFilePath(String tfID) throws IOException {
		return CdbFolders.getSubfolder(cdbParentFolder, CdbFolders.TF).resolve(transferFunsFileName);
	}
	
	/*
	 * (non-Javadoc)
	 * @see org.eso.ias.cdb.json.CdbFiles#getTemplateFilePath(java.lang.String)
	 */
	@Override
	public Path getTemplateFilePath(String templateID) throws IOException {
		return CdbFolders.getSubfolder(cdbParentFolder, CdbFolders.TEMPLATE).resolve(templatesFileName);
	}

	/**
	 *
	 * @param clientID The identifier of the client
	 * @return The path for the configuration of the client
	 *         with the passed ID
	 * @throws IOException In case of IO error getting the path
	 */
	@Override
	public Path getClientFilePath(String clientID) throws IOException {
		if (clientID==null || clientID.isEmpty()) {
			throw new IllegalArgumentException("Invalid null or empty DASU ID");
		}
		return CdbFolders.getSubfolder(cdbParentFolder, CdbFolders.CLIENT).resolve(clientID+confFileExtension);
	}

	/**
	 * @param clientID The identifier of the plugin
	 * @return The path for the configuration of the plugin
	 * with the passed ID
	 * @throws IOException In case of IO error getting the path
	 */
	@Override
	public Path getPluginFilePath(String clientID) throws IOException {
		if (clientID==null || clientID.isEmpty()) {
			throw new IllegalArgumentException("Invalid null or empty DASU ID");
		}
		return CdbFolders.getSubfolder(cdbParentFolder, CdbFolders.PLUGIN).resolve(clientID+ fileExtension);
	}
}
